### ISSUE TITLE
Loading pretrained chatbot_model1.pkl.gz fails in Colab NB

### ISSUE THREAD

**Issue by rickqiu** [google/trax#1080]
### Description
In Colab NB, it wouldn't be able to initialise a ReformerLM from pre-trained weights.
...

### Environment information

```
OS: 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.5 LTS"
NAME="Ubuntu"
VERSION="18.04.5 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.5 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic

$ pip freeze | grep trax
trax==1.3.4

$ pip freeze | grep tensor
mesh-tensorflow==0.1.16
tensor2tensor==1.15.7
tensorboard==2.3.0
tensorboard-plugin-wit==1.7.0
tensorboardcolab==0.0.22
tensorflow==2.3.0
tensorflow-addons==0.8.3
tensorflow-datasets==2.1.0
tensorflow-estimator==2.3.0
tensorflow-gan==2.0.0
tensorflow-gcs-config==2.3.0
tensorflow-hub==0.9.0
tensorflow-metadata==0.24.0
tensorflow-privacy==0.2.2
tensorflow-probability==0.7.0
tensorflow-text==2.3.0

$ pip freeze | grep jax
jax==0.2.0
jaxlib==0.1.55

$ python -V
Python 3.6.9

### For bugs: reproduction and error logs

```
# Steps to reproduce:
...

#### define the `predict_mem_len` and `predict_drop_len` of tl.SelfAttention
def attention(*args, **kwargs):
    # number of input positions to remember in a cache when doing fast inference. 
    kwargs['predict_mem_len'] = 120
    # number of input elements to drop once the fast inference input cache fills up.
    kwargs['predict_drop_len'] = 120
    # return the attention layer with the parameters defined above
    return tl.SelfAttention(*args, **kwargs)

#### define the model using the ReformerLM function you implemented earlier.
model = ReformerLM(
    vocab_size=33000,
    n_layers=6,
    mode='predict',
    attention_type=attention
)

#### define an input signature so we can initialize our model. shape will be (1, 1) and the data type is int32.
shape11 = trax.shapes.ShapeDtype((1, 1), dtype=np.int32)
#### initialize from file
model.init_from_file('/content/drive/My Drive/model/chatbot_model1.pkl.gz',
                     weights_only=True, input_signature=shape11)

#### save the starting state
STARTING_STATE = model.state

```
# Error logs:
LayerError                                Traceback (most recent call last)
<ipython-input-19-f3f97bdcafae> in <module>()
      1 # initialize from file
      2 model.init_from_file('/content/drive/My Drive/model/chatbot_model1.pkl.gz',
----> 3                      weights_only=True, input_signature=shape11)
      4 
      5 # save the starting state

1 frames
/usr/local/lib/python3.6/dist-packages/trax/layers/base.py in init(self, input_signature, rng, use_cache)
    286       name, trace = self._name, _short_traceback(skip=3)
    287       raise LayerError(name, 'init', self._caller,
--> 288                        input_signature, trace) from None
    289 
    290   def init_from_file(self, file_name, weights_only=False, input_signature=None):

LayerError: Exception passing through layer Serial (in init):
  layer created in file [...]/models/reformer/reformer.py, line 360
  layer input shapes: ShapeDtype{shape:(1, 1), dtype:<class 'numpy.int32'>}

  File [...]/trax/layers/combinators.py, line 104, in init_weights_and_state
    sublayer.init(inputs, use_cache=True))

LayerError: Exception passing through layer Embedding_33000_512 (in init):
  layer created in file [...]/models/reformer/reformer.py, line 326
  layer input shapes: ShapeDtype{shape:(1, 1), dtype:int32}

  File [...]/trax/layers/core.py, line 157, in init_weights_and_state
    w = self._kernel_initializer(shape_w, self.rng)

  File [...]/trax/layers/initializers.py, line 72, in <lambda>
    rng, _PureShape(shape)).astype('float32'))

  File [...]/trax/fastmath/ops.py, line 71, in normal
    return backend()['random_normal'](*args, **kwargs)

  File [...]/dist-packages/jax/random.py, line 634, in normal
    return _normal(key, shape, dtype)  # type: ignore

  File [...]/dist-packages/jax/api.py, line 215, in f_jitted
    donated_invars=donated_invars)

  File [...]/dist-packages/jax/core.py, line 1144, in bind
    return call_bind(self, fun, *args, **params)

  File [...]/dist-packages/jax/core.py, line 1133, in call_bind
    tracers = map(top_trace.full_raise, args)

  File [...]/dist-packages/jax/util.py, line 35, in safe_map
    return list(map(f, *args))

  File [...]/dist-packages/jax/core.py, line 359, in full_raise
    val._assert_live()

  File [...]/jax/interpreters/partial_eval.py, line 819, in _assert_live
    raise core.escaped_tracer_error(msg)

jax.core.UnexpectedTracerError: Encountered an unexpected tracer. Perhaps this tracer escaped through global state from a previously traced function.
The functions being transformed should not save traced values to global state. Detail: tracer created on line /usr/local/lib/python3.6/dist-packages/trax/fastmath/ops.py:59 (get_prng).
```

**Comment by lukaszkaiser** [comment]
I believe this is corrected in newer versions of Trax. Closing, but please re-open if you see a Tracer error in newer Trax.

**Comment by at-creator** [comment]
Reformer Text Generation notebook give same error while initializing with pretrained weights into the model : https://colab.research.google.com/github/google/trax/blob/master/trax/models/reformer/text_generation.ipynb







LayerError                                Traceback (most recent call last)
<ipython-input-13-3690a7a9cda7> in <module>()
      2 model = trax.models.ReformerLM(mode='predict')
      3 model.init_from_file(os.path.join(output_dir,'model.pkl.gz'),
----> 4                      weights_only=True)
      5 
      6 # Sample from ReformerLM

1 frames
/usr/local/lib/python3.6/dist-packages/trax/layers/base.py in init(self, input_signature, rng, use_cache)
    305       name, trace = self._name, _short_traceback(skip=3)
    306       raise LayerError(name, 'init', self._caller,
--> 307                        input_signature, trace) from None
    308 
    309   def init_from_file(self, file_name, weights_only=False, input_signature=None):

LayerError: Exception passing through layer Serial (in init):
  layer created in file [...]/models/reformer/reformer.py, line 185
  layer input shapes: (ShapeDtype{shape:(1, 524288), dtype:int32}, ShapeDtype{shape:(1, 524288), dtype:int32}, ShapeDtype{shape:(1, 524288), dtype:float32})

  File [...]/trax/layers/combinators.py, line 105, in init_weights_and_state
    sublayer.init(inputs, use_cache=True))

LayerError: Exception passing through layer ReversibleSerial (in init):
  layer created in file [...]/models/reformer/reformer.py, line 179
  layer input shapes: (ShapeDtype{shape:(1, 524288, 256), dtype:float32}, ShapeDtype{shape:(1, 524288, 256), dtype:float32})

  File [...]/trax/layers/combinators.py, line 105, in init_weights_and_state
    sublayer.init(inputs, use_cache=True))

LayerError: Exception passing through layer ReversibleHalfResidual (in init):
  layer created in file [...]/models/reformer/reformer.py, line 69
  layer input shapes: (ShapeDtype{shape:(1, 524288, 256), dtype:float32}, ShapeDtype{shape:(1, 524288, 256), dtype:float32})

  File [...]/trax/layers/reversible.py, line 326, in init_weights_and_state
    attn_weights, attn_state = self.attention_layer.init(inputs)

LayerError: Exception passing through layer SelfAttention (in init):
  layer created in file [...]/layers/research/efficient_attention.py, line 1009
  layer input shapes: ShapeDtype{shape:(1, 524288, 256), dtype:float32}

  File [...]/layers/research/efficient_attention.py, line 1072, in init_weights_and_state
    self.rng, weights_rng, state_rng = fastmath.random.split(self.rng, 3)

  File [...]/trax/fastmath/ops.py, line 71, in split
    return backend()['random_split'](prng, num)

  File [...]/jax/_src/random.py, line 259, in split
    return _split(key, int(num))  # type: ignore

  File [...]/dist-packages/jax/api.py, line 371, in f_jitted
    return cpp_jitted_f(*args, **kwargs)

  File [...]/dist-packages/jax/api.py, line 284, in cache_miss
    donated_invars=donated_invars)

  File [...]/dist-packages/jax/core.py, line 1229, in bind
    return call_bind(self, fun, *args, **params)

  File [...]/dist-packages/jax/core.py, line 1218, in call_bind
    tracers = map(top_trace.full_raise, args)

  File [...]/dist-packages/jax/util.py, line 36, in safe_map
    return list(map(f, *args))

  File [...]/dist-packages/jax/core.py, line 378, in full_raise
    val._assert_live()

  File [...]/jax/interpreters/partial_eval.py, line 950, in _assert_live
    raise core.escaped_tracer_error(msg)

jax.core.UnexpectedTracerError: Encountered an unexpected tracer. Perhaps this tracer escaped through global state from a previously traced function.
The functions being transformed should not save traced values to global state. Detail: tracer created on line /usr/local/lib/python3.6/dist-packages/trax/layers/research/efficient_attention.py:1072 (init_weights_and_state).

**Comment by darkmatter999** [comment]
@at-creator 
https://colab.research.google.com/github/google/trax/blob/master/trax/models/reformer/text_generation.ipynb
raises above described error in the latest trax===1.3.7, but works in trax===1.3.6.
So the issue solved after trax===1.3.4 must have slipped in somehow again in 1.3.7.
